dashboard/lib/user_permission_concern.rb

```diff
@@ -0,0 +1,98 @@
+require 'cdo/chat_client'
+
+#
+# Module to mix-in to User model providing expressive interactions with
+# user permissions.
+#
+module UserPermissionConcern
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :permissions, class_name: 'UserPermission', dependent: :destroy
+    before_save :log_admin_save, if: -> {admin_changed? && UserPermissionConcern.should_log?}
+  end
+
+  # @param permission [UserPermission] the permission to query.
+  # @return [Boolean] whether the User has permission granted.
+  def permission?(permission)
+    return false unless teacher?
+    if @permissions.nil?
+      # The user's permissions have not yet been cached, so do the DB query,
+      # caching the results.
+      @permissions = UserPermission.where(user_id: id).pluck(:permission)
+    end
+    # Return the cached results.
+    @permissions.include? permission
+  end
+
+  def permission=(permission)
+    @permissions = nil
+    permissions << permissions.find_or_create_by(user_id: id, permission: permission)
+  end
+
+  def delete_permission(permission)
+    @permissions = nil
+    permission = permissions.find_by(permission: permission)
+    permissions.delete permission if permission
+  end
+
+  # Revokes all escalated permissions associated with the user, including admin status and any
+  # granted UserPermission's.
+  def revoke_all_permissions
+    update_column(:admin, nil)
+    UserPermission.where(user_id: id).each(&:destroy)
```

Comment:
nit: this should also set `@permissions` to nil

Reply:
Thank you for catching this!  I'd like to leave this PR as a simple extraction of the methods in question, and make fixes/improvements in a follow-up.  Sound okay?